# Data Cleaning & Preparation: NYC Airbnb 2019
Dataset: AB_NYC_2019 (48,895 listings, 16 columns)
Goal: find and fix missing values, duplicates, wrong data types and inconsistent values.

In [7]:
import pandas as pd
import numpy as np

In [8]:
df_raw = pd.read_csv(r"C:\Users\manso\OneDrive\Documents\SWYNEX-Data-Cleaning-Preparation\Data\AB_NYC_2019.csv")
df_raw

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaN,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48890,36484665,Charming one bedroom - newly renovated rowhouse,8232441,Sabrina,Brooklyn,Bedford-Stuyvesant,40.67853,-73.94995,Private room,70,2,0,NaN,NaN,2,9
48891,36485057,Affordable room in Bushwick/East Williamsburg,6570630,Marisol,Brooklyn,Bushwick,40.70184,-73.93317,Private room,40,4,0,NaN,NaN,2,36
48892,36485431,Sunny Studio at Historical Neighborhood,23492952,Ilgar & Aysel,Manhattan,Harlem,40.81475,-73.94867,Entire home/apt,115,10,0,NaN,NaN,1,27
48893,36485609,43rd St. Time Square-cozy single bed,30985759,Taz,Manhattan,Hell's Kitchen,40.75751,-73.99112,Shared room,55,1,0,NaN,NaN,6,2


In [9]:
df = df_raw.copy()
df.shape

(48895, 16)

In [5]:
df.head()

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaN,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0


In [13]:
df.info()
#last_review is stored as text and needs to be converted to a date.

<class 'pandas.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  str    
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  str    
 4   neighbourhood_group             48895 non-null  str    
 5   neighbourhood                   48895 non-null  str    
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  str    
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     38843 non-n

In [12]:
df.isnull().sum()
#4 columns have missing values.

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [14]:
df.describe()
#price has a minimum of 0, which needs checking.

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.889500e+04,4.889500e+04,48895.000000,48895.000000,48895.000000,48895.000000,48895.000000,38843.000000,48895.000000,48895.000000
mean,1.901714e+07,6.762001e+07,40.728949,-73.952170,152.720687,7.029962,23.274466,1.373221,7.143982,112.781327
std,1.098311e+07,7.861097e+07,0.054530,0.046157,240.154170,20.510550,44.550582,1.680442,32.952519,131.622289
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,0.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.471945e+06,7.822033e+06,40.690100,-73.983070,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.967728e+07,3.079382e+07,40.723070,-73.955680,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.915218e+07,1.074344e+08,40.763115,-73.936275,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


In [17]:
before = {
    "rows": df_raw.shape[0],
    "columns": df_raw.shape[1],
    "missing_total": int(df_raw.isnull().sum().sum()),
    "duplicates": int(df_raw.duplicated().sum()),
}
before

{'rows': 48895, 'columns': 16, 'missing_total': 20141, 'duplicates': 0}

In [24]:
df['reviews_per_month'] = df['reviews_per_month'].fillna(0)

In [25]:
df['name'] = df['name'].fillna("Unknown")
df['host_name'] = df['host_name'].fillna("Unknown")

Filled reviews_per_month with 0 for listings with no reviews (not truly missing — 0 reviews means 0 per month). Filled name and host_name with "Unknown" (37 rows total). Left last_review blank for listings with no reviews, since there is no valid date to assign.

In [26]:
df.isnull().sum()

id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                     0
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [27]:
dupes = df[df.duplicated(subset=['host_id', 'name'], keep=False)]
dupes.sort_values(['host_id', 'name']).head(20)

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
73,17693,"HARLEM, NEW YORK WELCOMES YOU!!",68428,Tye And Etienne,Manhattan,Washington Heights,40.83139,-73.94095,Private room,65,2,49,2019-06-18,1.60,2,336
3581,2138787,"HARLEM, NEW YORK WELCOMES YOU!!",68428,Tye And Etienne,Manhattan,Washington Heights,40.83191,-73.94110,Private room,62,2,4,2019-04-01,0.19,2,336
1066,449034,Spacious Loft 5 min to Union Square,124357,Alex,Brooklyn,Williamsburg,40.71756,-73.95282,Private room,90,1,155,2019-06-07,1.78,3,328
2564,1330056,Spacious Loft 5 min to Union Square,124357,Alex,Brooklyn,Williamsburg,40.71571,-73.95813,Private room,90,1,115,2019-05-21,1.63,3,359
327,80684,Duplex w/ Terrace @ Box House Hotel,417504,The Box House Hotel,Brooklyn,Greenpoint,40.73776,-73.95327,Private room,349,3,8,2016-03-27,0.09,28,60
679,253828,Duplex w/ Terrace @ Box House Hotel,417504,The Box House Hotel,Brooklyn,Greenpoint,40.73674,-73.95247,Private room,349,3,8,2018-07-26,0.09,28,58
652,248865,Loft Suite @ Box House Hotel,417504,The Box House Hotel,Brooklyn,Greenpoint,40.73787,-73.95385,Entire home/apt,199,3,34,2019-04-22,0.37,28,60
674,253800,Loft Suite @ Box House Hotel,417504,The Box House Hotel,Brooklyn,Greenpoint,40.73730,-73.95323,Entire home/apt,199,3,24,2019-04-25,0.26,28,60
831,294259,Loft Suite @ Box House Hotel,417504,The Box House Hotel,Brooklyn,Greenpoint,40.73806,-73.95462,Entire home/apt,199,3,73,2019-06-30,0.84,28,56
836,294490,Loft Suite @ Box House Hotel,417504,The Box House Hotel,Brooklyn,Greenpoint,40.73857,-73.95299,Entire home/apt,199,3,54,2019-06-10,0.61,28,60


In [28]:
near_dupes_count = df.duplicated(subset=['host_id', 'name'], keep=False).sum()
print(f"Found {near_dupes_count} rows sharing host_id + name — inspected and confirmed these are distinct units (different id, lat/long, review history), not duplicates. No rows dropped.")

Found 410 rows sharing host_id + name — inspected and confirmed these are distinct units (different id, lat/long, review history), not duplicates. No rows dropped.


Checked for duplicates two ways: (1) exact duplicate rows → 0 found. (2) Same host_id + name → 243 rows found, but inspection showed each has a unique id, different coordinates, and different review counts — these are separate rooms with reused listing names (e.g., hotels with many similarly-named units), not duplicates. No rows were removed at this stage.

In [29]:
df['last_review'] = pd.to_datetime(df['last_review'])
df['neighbourhood_group'] = df['neighbourhood_group'].astype('category')
df['neighbourhood'] = df['neighbourhood'].astype('category')
df['room_type'] = df['room_type'].astype('category')

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48895 non-null  int64         
 1   name                            48895 non-null  str           
 2   host_id                         48895 non-null  int64         
 3   host_name                       48895 non-null  str           
 4   neighbourhood_group             48895 non-null  category      
 5   neighbourhood                   48895 non-null  category      
 6   latitude                        48895 non-null  float64       
 7   longitude                       48895 non-null  float64       
 8   room_type                       48895 non-null  category      
 9   price                           48895 non-null  int64         
 10  minimum_nights                  48895 non-null  int64         
 11  number_of_rev

In [30]:
df['name'] = df['name'].str.strip()

In [31]:
(df['name'] != df['name'].str.strip()).sum()

np.int64(0)

Removed extra spaces from the start and end of listing names (238 rows affected). Left the capitalization style as-is since it reflects intentional host choices, not a data error.

In [32]:
df = df[df['price'] > 0]

In [33]:
df = df[df['minimum_nights'] <= 365]

In [34]:
df.shape

(48870, 16)

Removed 11 listings with price == 0 (invalid — Airbnb listings cannot be free) and 14 listings with minimum_nights over 365 (unrealistic booking requirement). Rows dropped from 48,895 to 48,870.

In [35]:
df.to_csv("data/AB_NYC_2019_cleaned.csv", index=False)

In [36]:
after = {
    "rows": df.shape[0],
    "columns": df.shape[1],
    "missing_total": int(df.isnull().sum().sum()),
    "duplicates": int(df.duplicated().sum()),
}
after

{'rows': 48870, 'columns': 16, 'missing_total': 10043, 'duplicates': 0}

In [37]:
import pandas as pd
comparison = pd.DataFrame([before, after], index=['Before', 'After'])
comparison

,rows,columns,missing_total,duplicates
Before,48895,16,20141,0
After,48870,16,10043,0
